core/sr-io/without_std.rs

```diff
@@ -624,10 +718,200 @@ impl CryptoApi for () {
 }
 
 impl OffchainApi for () {
-	fn submit_extrinsic<T: codec::Encode>(data: &T) {
+	fn submit_transaction<T: codec::Encode>(data: &T) -> Result<(), ()> {
 		let encoded_data = codec::Encode::encode(data);
 		unsafe {
-			ext_submit_extrinsic.get()(encoded_data.as_ptr(), encoded_data.len() as u32)
+			let ret = ext_submit_transaction.get()(encoded_data.as_ptr(), encoded_data.len() as u32);
+			if ret == 0 {
+				Ok(())
+			} else {
+				Err(())
+			}
+		}
+	}
+
+	fn sign(data: &[u8]) -> Option<[u8; 64]> {
+		let mut result = [0_u8; 64];
+		unsafe {
+			let res = ext_sign.get()(data.as_ptr(), data.len() as u32, result.as_mut_ptr());
+			if res == 0 {
+				Some(result)
+			} else {
+				None
+			}
+		}
+	}
+
+	fn timestamp() -> offchain::Timestamp {
+		offchain::Timestamp::from_unix_millis(unsafe {
+			ext_timestamp.get()()
+		})
+	}
+
+	fn sleep_until(deadline: Timestamp) {
+		unsafe {
+			ext_sleep_until.get()(deadline.unix_millis())
+		}
+	}
+
+	fn random_seed() -> [u8; 32] {
+		let mut result = [0_u8; 32];
+		unsafe {
+			ext_random_seed.get()(result.as_mut_ptr())
+		}
+		result
+	}
+
+	fn local_storage_set(key: &[u8], value: &[u8]) {
+		unsafe {
+			ext_local_storage_set.get()(
+				key.as_ptr(),
+				key.len() as u32,
+				value.as_ptr(),
+				value.len() as u32,
+			)
+		}
+	}
+
+	fn local_storage_read(key: &[u8]) -> Option<Vec<u8>> {
+		let mut len = 0u32;
+		unsafe {
+			let ptr = ext_local_storage_read.get()(
+				key.as_ptr(),
+				key.len() as u32,
+				&mut len,
+			);
+
+			if len == u32::max_value() {
+				None
+			} else {
+
+				// Invariants required by Vec::from_raw_parts are not formally fulfilled.
+				// We don't allocate via String/Vec<T>, but use a custom allocator instead.
+				// See #300 for more details.
+				Some(<Vec<u8>>::from_raw_parts(ptr, len as usize, len as usize))
+			}
+		}
+	}
+
+	fn http_request_start(method: &str, url: &str, meta: &[u8]) -> Result<offchain::HttpRequestId, ()> {
+		let method = method.as_bytes();
+		let url = url.as_bytes();
+
+		unsafe {
+			let result = ext_http_request_start.get()(
+				method.as_ptr(),
+				method.len() as u8,
```

Comment:
Although truncation here cannot lead to unsafety, maybe it would still be good to return `Err` if the conversion to `u8` fails?

Reply:
Changed to `u32`